File: back/src/main/java/com/back/domain/concert/service/ConcertService.java
package com.back.domain.concert.service;

import com.back.domain.concert.dto.ConcertDetailResponse;
import com.back.domain.concert.dto.ConcertListResponse;
import com.back.domain.concert.entity.Concert;
import com.back.domain.concert.entity.ConcertDetail;
import com.back.domain.concert.enums.ConcertSortType;
import com.back.domain.concert.repository.ConcertDeatilRepository;
import com.back.domain.concert.repository.ConcertRepository;
import com.back.domain.schedule.entity.Schedule;
import com.back.domain.schedule.entity.ScheduleSeat;
import com.back.domain.schedule.entity.SeatStatus;
import com.back.domain.schedule.repository.ScheduleRepository;
import com.back.domain.schedule.repository.ScheduleSeatRepository;
import com.back.global.exception.ErrorCode;
import com.back.global.exception.ServiceException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class ConcertService {
    private final ScheduleSeatRepository scheduleSeatRepository;
    private final ScheduleRepository scheduleRepository;
    private final ConcertRepository concertRepository;
    private final ConcertDeatilRepository concertDeatilRepository;

    public List<ConcertListResponse> getConcerts(String keyword, ConcertSortType sort) {
        List<Concert> concerts = concertRepository.findByKeyword(keyword);

        List<Long> concertIds = concerts.stream()
                .map(Concert::getConcertId)
                .toList();

        List<Schedule> schedules = scheduleRepository.findAllWithVenueByConcertIds(concertIds);

        Map<Long, String> venueNameMap = schedules.stream()
                .collect(Collectors.toMap(
                        schedule -> schedule.getConcert().getConcertId(),
                        schedule -> schedule.getVenue().getVenueName(),
                        (existing, replacement) -> existing
                ));

        Comparator<Concert> comparator;
        if (sort == ConcertSortType.latest) {
            comparator = Comparator.comparing(Concert::getStartDate).reversed();
        } else {
            comparator = Comparator.comparing(Concert::getEndDate);
        }

        return concerts.stream()
                .sorted(comparator)
                .map(concert -> {
                    String venueName = venueNameMap.getOrDefault(concert.getConcertId(), "");
                    return ConcertListResponse.of(concert, venueName);
                })
                .collect(Collectors.toList());
    }

    public ConcertDetailResponse getConcertDetail(Long concertId) {
        Concert concert = concertRepository.findById(concertId)
                .orElseThrow(() -> new ServiceException(ErrorCode.CONCERT_NOT_FOUND));

        Schedule schedule = scheduleRepository.findWithVenueByConcertId(concertId)
                .stream()
                .findFirst()
                .orElseThrow(() -> new ServiceException(ErrorCode.CONCERT_SCHEDULE_EMPTY));

        List<String> detailUrlList = concertDeatilRepository
                .findByConcertConcertId(concertId)
                .stream()
                .map(ConcertDetail::getUrlDetail)
                .collect(Collectors.toList());

        List<ScheduleSeat> scheduleSeats = scheduleSeatRepository.findByScheduleScheduleId(schedule.getScheduleId());
        Map<String, Integer> prices = convertToPriceMap(scheduleSeats);

        boolean bookable = concert.getEndDate().isAfter(LocalDateTime.now());

        return ConcertDetailResponse.of(
                concert,
                schedule.getVenue().getVenueName(),
                schedule.getVenue().getLocation(),
                detailUrlList,
                prices,
                bookable
        );
    }

    public List<ScheduleSeat> getScheduleSeats(Long scheduleId) {
        return scheduleSeatRepository.findByScheduleScheduleId(scheduleId);
    }

    public void validateSeatAvailable(Long scheduleId, String seatNumber) {
        ScheduleSeat seat = scheduleSeatRepository
                .findWithLockByScheduleIdAndSeatNumber(scheduleId, seatNumber)
                .orElseThrow(() -> new ServiceException(ErrorCode.SEAT_NOT_FOUND));

        if (seat.getSeatStatus() == SeatStatus.SOLD_OUT) {
            throw new ServiceException(ErrorCode.SEAT_ALREADY_SOLD);
        }
    }

    public void validateConcertScheduleMatch(Long concertId, Long scheduleId) {
        scheduleRepository.findByScheduleIdAndConcert_ConcertId(scheduleId, concertId)
                .orElseThrow(() -> new ServiceException(ErrorCode.INVALID_CONCERT_SCHEDULE));
    }

    public Map<String, Integer> convertToPriceMap(List<ScheduleSeat> scheduleSeats) {
        return scheduleSeats.stream()
                .collect(Collectors.toMap(
                        ScheduleSeat::getGradeName,
                        ScheduleSeat::getSeatPrice,
                        (oldPrice, newPrice) -> oldPrice
                ));
    }
}

File: back/src/main/kotlin/com/back/domain/concert/service/ConcertService.kt
package com.back.domain.concert.service

import com.back.domain.concert.dto.ConcertDetailResponse
import com.back.domain.concert.dto.ConcertListResponse
import com.back.domain.concert.entity.Concert
import com.back.domain.concert.constant.ConcertSortType
import com.back.domain.concert.repository.ConcertDeatilRepository
import com.back.domain.concert.repository.ConcertRepository
import com.back.domain.schedule.entity.ScheduleSeat
import com.back.domain.schedule.constant.SeatStatus
import com.back.domain.schedule.repository.ScheduleRepository
import com.back.domain.schedule.repository.ScheduleSeatRepository
import com.back.global.exception.ErrorCode
import com.back.global.exception.ServiceException
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDate
import java.time.LocalDateTime

@Service
@Transactional(readOnly = true)
class ConcertService(
    private val scheduleSeatRepository: ScheduleSeatRepository,
    private val scheduleRepository: ScheduleRepository,
    private val concertRepository: ConcertRepository,
    private val concertDeatilRepository: ConcertDeatilRepository
) {

    fun getConcerts(keyword: String?, sort: ConcertSortType?, date: LocalDate?): List<ConcertListResponse> {
        val concerts = concertRepository.findByKeyword(keyword)
        val concertIds = concerts.mapNotNull { it.concertId }
        val schedules = scheduleRepository.findAllWithVenueByConcertIds(concertIds)

        val filteredConcerts = date?.let { targetDate ->
            val matchingIds = schedules
                .filter { it.scheduleDate.toLocalDate() == targetDate }
                .mapNotNull { it.concert.concertId }
                .toSet()
            concerts.filter { matchingIds.contains(it.concertId) }
        } ?: concerts

        val venueNameMap = schedules
            .mapNotNull { s -> s.concert.concertId?.let { cid -> cid to s.venue.venueName } }
            .toMap()

        val today = LocalDate.now().atStartOfDay()

        val sortTarget = if (sort == ConcertSortType.closingSoon) {
            filteredConcerts.filter { it.endDate >= today }
        } else {
            filteredConcerts
        }

        val comparator = when (sort) {
            ConcertSortType.latest -> compareByDescending<Concert> { it.concertId }
            else -> compareBy<Concert> { it.endDate }
        }

        return sortTarget.sortedWith(comparator)
            .map { concert ->
                val venueName = venueNameMap[concert.concertId].orEmpty()
                ConcertListResponse.of(concert, venueName)
            }
    }

    fun getConcertDetail(concertId: Long): ConcertDetailResponse {
        val concert = concertRepository.findByIdOrNull(concertId)
            ?: throw ServiceException(ErrorCode.CONCERT_NOT_FOUND)

        val schedule = scheduleRepository.findWithVenueByConcertId(concertId)
            .firstOrNull()
            ?: throw ServiceException(ErrorCode.CONCERT_SCHEDULE_EMPTY)

        val detailUrlList = concertDeatilRepository
            .findByConcertConcertId(concertId)
            .mapNotNull { it.urlDetail }

        val scheduleId = checkNotNull(schedule.scheduleId) { "Schedule ID null" }
        val scheduleSeats = scheduleSeatRepository.findByScheduleScheduleId(scheduleId)
        val prices = convertToPriceMap(scheduleSeats)
        val bookable = concert.isBookable()

        return ConcertDetailResponse.of(
            concert = concert,
            venueName = schedule.venue.venueName,
            location = schedule.venue.location,
            detailUrlList = detailUrlList,
            prices = prices,
            bookable = bookable
        )
    }

    fun getScheduleSeats(scheduleId: Long): List<ScheduleSeat> =
        scheduleSeatRepository.findByScheduleScheduleId(scheduleId)

    @Transactional
    fun validateSeatAvailable(scheduleId: Long, seatNumber: String) {
        val seat = scheduleSeatRepository
            .findWithLockByScheduleIdAndSeatNumber(scheduleId, seatNumber)
            ?: throw ServiceException(ErrorCode.SEAT_NOT_FOUND)

        when (seat.seatStatus) {
            SeatStatus.SOLD_OUT -> throw ServiceException(ErrorCode.SEAT_ALREADY_SOLD)
            SeatStatus.HOLD -> throw ServiceException(ErrorCode.SEAT_HELD_BY_OTHER_USER)
            else -> Unit
        }
    }

    fun validateConcertScheduleMatch(concertId: Long, scheduleId: Long) {
        scheduleRepository.findByScheduleIdAndConcert_ConcertId(scheduleId, concertId)
            ?: throw ServiceException(ErrorCode.INVALID_CONCERT_SCHEDULE)
    }

    fun convertToPriceMap(scheduleSeats: List<ScheduleSeat>): Map<String, Int> =
        scheduleSeats.associate { it.gradeName to it.seatPrice }

    fun validateScheduleBookable(scheduleId: Long) {
        val schedule = scheduleRepository.findByIdOrNull(scheduleId)
            ?: throw ServiceException(ErrorCode.CONCERT_SCHEDULE_EMPTY)

        if (schedule.isExpired()) {
            throw ServiceException(ErrorCode.EXPIRED_BOOKING_DEADLINE)
        }
    }
}
